src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Flow.java

```diff
@@ -740,220 +732,452 @@ public void visitSwitchExpression(JCSwitchExpression tree) {
+                            .map(r -> types.memberType(record.type, r))
+                            .toArray(s -> new Type[s]);
+                    PatternDescription[] nestedDescriptions =
+                            new PatternDescription[record.nested.size()];
+                    int i = 0;
+                    for (List<JCPattern> it = record.nested;
+                         it.nonEmpty();
+                         it = it.tail, i++) {
+                        nestedDescriptions[i] = PatternDescription.from(types, componentTypes[i], it.head);
                     }
-                    case RECORDPATTERN -> {
-                        JCRecordPattern dpat = (JCRecordPattern) labelValue;
-                        UniqueType type = new UniqueType(dpat.type, types);
-                        List<JCRecordPattern> augmentedPatterns =
-                                deconstructionPatternsByType.getOrDefault(type, List.nil())
-                                                                 .prepend(dpat);
-
-                        deconstructionPatternsByType.put(type, augmentedPatterns);
+                    return new RecordPattern(record.type, componentTypes, nestedDescriptions);
+                } else {
+                    throw Assert.error();
+                }
+            }
+        }
+
+        record BindingPattern(Type type) implements PatternDescription {
+            @Override
+            public int hashCode() {
+                return type.tsym.hashCode();
+            }
+            @Override
+            public boolean equals(Object o) {
+                return o instanceof BindingPattern other &&
+                       type.tsym == other.type.tsym;
+            }
+            @Override
+            public String toString() {
+                return type.tsym + " _";
+            }
+        }
+
+        record RecordPattern(Type recordType, int _hashCode, Type[] fullComponentTypes, PatternDescription... nested) implements PatternDescription {
+
+            public RecordPattern(Type recordType, Type[] fullComponentTypes, PatternDescription[] nested) {
+                this(recordType, hashCode(-1, recordType, fullComponentTypes, nested), fullComponentTypes, nested);
+            }
+
+            @Override
+            public int hashCode() {
+                return _hashCode;
+            }
+
+            @Override
+            public boolean equals(Object o) {
+                return o instanceof RecordPattern other &&
+                       recordType.tsym == other.recordType.tsym &&
+                       Arrays.equals(nested, other.nested);
+            }
+
+            public int hashCode(int excludeComponent) {
+                return hashCode(excludeComponent, recordType, fullComponentTypes, nested);
+            }
+
+            public static int hashCode(int excludeComponent, Type recordType, Type[] fullComponentTypes, PatternDescription... nested) {
```

Comment:
this argument: `fullComponentTypes` is not being used in this method, could be removed